src/clj/puppetlabs/services/jruby/jruby_metrics_core.clj

```diff
@@ -152,6 +166,73 @@
    (comp add-duration-to-instance instance-request-info)
    instances))
 
+(schema/defn timer-for-borrow-reason :- Timer
+  "Create or return a named timer for a JRuby borrow reason.
+
+  If the JRuby instance was borrowed to service a HTTP request, then
+  the timer name is generated from the sanitized Comidi :route-id.
+
+  If a Clojure Keyword was used as the JRuby borrow reason, then the
+  timer is generated from the Keyword's name and namespace.
+
+  Returns a named labeled 'other' if the borrow reason does not match
+  one of the above cases."
+  [{:keys [hostname metric-registry]} :- JRubyMetrics
+   request :- TimestampedReason]
+  (let [request-info (instance-request-info request)
+        reason (:reason request-info)
+        timer-name (cond
+                     (nil? (schema/check HttpRequestReasonInfo reason))
+                       (-> (get-in reason [:request :route-id])
+                           ;; Remove ":" characters created by stringifying
+                           ;; Clojure keywords.
+                           (str/replace ":" "")
+                           ;; Remove segments created from regexes by Comidi.
+                           ;; Regexes contain characters that create escaping
+                           ;; problems in downstream services that use regexes
+                           ;; to search for or select metrics by name.
+                           (str/replace #"-\/.*\/" ""))
+                     (keyword? reason)
+                       (if-let [name-space (namespace reason)]
+                         (str name-space "." (name reason))
+                         (name reason))
+                     :else "other")]
+    (->> timer-name
+         (str "jruby.borrow-timer.")
+         (metrics/host-metric-name hostname)
+         (.timer metric-registry))))
```

Comment:
Do we want to remove older metrics when creating newer ones? Is there a hard stop at how many borrow metrics we may have?

Reply:
Right now there is no hard stop. We just dynamically make a new counter for each borrow reason. The MetricRegistry class takes care of ensuring this is done in a thread-safe manner.

We could implement some sort of cleanup, but we already have MetricsPuppetProfiler.java dynamically creating a separate timer for each unique combination of certname, environment, and resource or function name that occurs during catalog compilation. That commonly creates 1000s of timer objects, so we don't have much to gain by implementing cleanup logic for the couple dozen that would be created here.

We should also totally look into the RAM usage from spawning all those timers in MetricsPuppetProfiler.java. I don't think it has ever presented as a "memory leak", but it is an unbounded allocator of a lot of persistent objects.